Add Swerve

dslit3 opened the pull request:
From [https://github.com/momentumfrc/swerve-chassis/](https://github.com/momentumfrc/swerve-chassis/)

Currently broken:
- Positioning
- Path planning, all of autonomous
- Field oriented (probably)
- Anything relying on differential/holonomic drive

None of this code has been tested

jordan-powers committed:
Add LEDSubsystem

jordan-powers committed:
Show limelight crosshairs alignment on LEDs

momentumrobotics-4999 committed:
Revert "Show limelight crosshairs alignment on LEDs"

This reverts commit 61d086c78bece110ea1a36bd9e012cbbfa97b86a.

momentumrobotics-4999 committed:
Merge branch 'main' into leds

momentumrobotics-4999 committed:
Use AddressableLEDDisplay

dslit3 committed:
Integrate Swerve from https://github.com/momentumfrc/swerve-chassis/

Motor speeds are most likely completely wrong
Positioning subsystem does not work, and by extension path planning
Field oriented drive is likely broken
Differential drive does not work
Lots of non-working or otherwise vestigial code (See all TODO)

dslit3 committed:
Absolute encoder fix

momentumrobotics-4999 committed:
Remove extra argument in CenterLimelightCrosshairsCommand

momentumrobotics-4999 committed:
Merge remote-tracking branch 'origin/swerve' into swerve

momentumrobotics-4999 committed:
Remove brakes

momentumrobotics-4999 committed:
Brake drive motors

momentumrobotics-4999 committed:
Made lots of changes

No commit message today -- I just want to go home.

jordan-powers commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -13,12 +14,29 @@
 
 public final class MoPrefs {
     public static Pref<Double> maxDriveRpm = doublePref("Drive Max RPM", 6000.0);
+    public static Pref<Double> maxDriveSpeed = doublePref("Drive Max Speed m/s", 6000.0);
```
I think these two preferences are mutually exclusive. That is, since we have maxDriveSpeed and driveMtrScale, we don't need maxDriveRpm.

I think maxDriveRpm was what we were using for the old mecanum drive pid, but we don't need it anymore for the swerve pid.

jordan-powers commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -13,12 +14,29 @@
     public static Pref<Double> turnSlowSpeed = doublePref("Turn Slow Speed", 0.25);
     public static Pref<Double> driveRampTime = doublePref("Drive Ramp Time", 0.25);
 
+    public static Pref<Double> flZero = doublePref("FL Drive Zero", 0);
+    public static Pref<Double> flScale = doublePref("FL Drive Scale", 1);
+    public static Pref<Double> frZero = doublePref("FR Drive Zero", 0);
+    public static Pref<Double> frScale = doublePref("FR Drive Scale", 1);
+    public static Pref<Double> rlZero = doublePref("RL Drive Zero", 0);
+    public static Pref<Double> rlScale = doublePref("RL Drive Scale", 1);
+    public static Pref<Double> rrZero = doublePref("RR Drive Zero", 0);
+    public static Pref<Double> rrScale = doublePref("RR Drive Scale", 1);
+
+    // Motor Scale: units of ticks per meter
+    public static Pref<Double> flDriveMtrScale = doublePref("FL Drive Motor Scale", 1);
+    public static Pref<Double> frDriveMtrScale = doublePref("FR Drive Motor Scale", 1);
+    public static Pref<Double> rlDriveMtrScale = doublePref("RL Drive Motor Scale", 1);
+    public static Pref<Double> rrDriveMtrScale = doublePref("RR Drive Motor Scale", 1);
```
We should probably adjust these names to be more obvious. For example, the difference between flScale and flDriveMtrScale is confusing, and flScale should probably be renamed something like flTurnMtrScale.

jordan-powers committed:
Use new motune implementation

jordan-powers committed:
Use motuner pid controller wrappers

jordan-powers committed:
Add motune vendordep

momentumrobotics-4999 committed:
Serve updates

momentumrobotics-4999 committed:
Revert "Use motuner pid controller wrappers"

This reverts commit a9dd6424cb6fbbe183840dfdf2b9f7d68ac32e2c.

momentumrobotics-4999 committed:
Hold heading

momentumrobotics-4999 committed:
Meeting changes

jordan-powers committed:
Update wpilib version

jordan-powers committed:
Improve swerve module re-zeroing logic

We really only want to re-zero the relative encoders on the
swerve modules if the following are true:
  1. The current encoder reading is valid (the filtered absolute encoder
     reading has not been invalidated by an active move or by
     electromagnetic interference from any nearby motors)
  2. We have already seen NUM_MEAN_SAMPLES consecutive valid encoder
     readings (we take the mean of several samples to account for the
     inherent variance of reading an analog value).
  3. It has been at least RESET_ENCODER_INTERVAL seconds since the last
     time the encoders were zeroed.
  4. Encoder zeroing has not been disabled.

This patch adds a dedicated class, FusedEncoders, to calculate these
four conditions and reset a relative encoder from a smoothed absolute
encoder reading if the conditions pass.

jordan-powers committed:
Oops, typos

jordan-powers committed:
Initial swerve path following

jordan-powers committed:
Curve/Deadzone moverequest magnitude
